### PR TITLE
lib.warn: Use or behave like builtins.warn

### PR DESCRIPTION
**Pull request by roberth**
## Description of changes

Base `lib.warn` on [`builtins.warn`](https://github.com/NixOS/nix/pull/10592) when available.
Behave like it when it's not available.

> A variation of `trace`, but
> 
>   * with the correct log level
> 
>   * no ANSI codes in expressions
> 
>   * can be debugged without stopping for actual `trace` calls (ie intentional "info level" or trace logging).
> 
>   * logs warnings without `trace:` in front of it
> 
>   * follows the Nix color scheme.

Alternative to, and
- Closes https://github.com/NixOS/nixpkgs/pull/306026

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: lib/trivial.nix [hunk 1]
```diff
@@ -698,9 +701,18 @@ in {
...
+    builtins.warn or (
+      let mustAbort = lib.elem (builtins.getEnv "NIX_ABORT_ON_WARN") ["1" "true" "yes"];
+      in
+        # Do not eta reduce v, so that we have the same strictness as `builtins.warn`.
+        msg: v:
+          # `builtins.warn` requires a string message, so we enforce that in our implementation, so that callers aren't accidentally incompatible with newer Nix versions.
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mwarning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
+          else builtins.trace "[1;31mwarning: ${msg}[0m" v
```
Comment (lorenzleutgeb) [inline lib/trivial.nix:714]:
```suggestion
          then builtins.trace "[1;31mevaluation warning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
          else builtins.trace "[1;31mevaluation warning: ${msg}[0m" v
```

compare https://github.com/NixOS/nix/pull/10592/files#diff-b3d97fa793934fc7b5faf1ae5e03c8db9fa28cdc4e0a61c47a08785e26805361R243-R244

**Comment by Aleksanaa** [comment]
image

The color seems not the same?

File: lib/trivial.nix [hunk 2]
```diff
@@ -698,9 +701,18 @@ in {
...
+    builtins.warn or (
+      let mustAbort = lib.elem (builtins.getEnv "NIX_ABORT_ON_WARN") ["1" "true" "yes"];
+      in
+        # Do not eta reduce v, so that we have the same strictness as `builtins.warn`.
+        msg: v:
+          # `builtins.warn` requires a string message, so we enforce that in our implementation, so that callers aren't accidentally incompatible with newer Nix versions.
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mevaluation warning: ${msg}[0m" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
+          else builtins.trace "[1;31mevaluation warning: ${msg}[0m" v
```
Comment (Aleksanaa) [inline lib/trivial.nix:714]:
```suggestion
          else builtins.trace "[1;35mevaluation warning: ${msg}[0m" v
```

To keep the color the same? But I don’t know why. It seems that Nix doesn’t like the escape sequence in the middle of the output string? So the following msg cannot be output in the default color.

Comment (Aleksanaa) [reply lib/trivial.nix:714]: Oh, there's a hidden character, my bad

**Comment by Aleksanaa** [comment]
After editing it to 35 and moving the hidden character (`\x1b`? Just copied it, forgot how to input it) in front of `${msg}`

Old nix:

image

New nix:

image

File: lib/trivial.nix [hunk 3]
```diff
@@ -698,9 +701,18 @@ in {
...
+    # Since https://github.com/NixOS/nix/pull/10592
+    builtins.warn or (
+      let mustAbort = lib.elem (builtins.getEnv "NIX_ABORT_ON_WARN") ["1" "true" "yes"];
+      in
+        # Do not eta reduce v, so that we have the same strictness as `builtins.warn`.
+        msg: v:
+          # `builtins.warn` requires a string message, so we enforce that in our implementation, so that callers aren't accidentally incompatible with newer Nix versions.
+          assert isString msg;
+          if mustAbort
+          then builtins.trace "[1;31mevaluation warning:[0m ${msg}" (abort "NIX_ABORT_ON_WARN=true; warnings are treated as unrecoverable errors.")
```
Comment (roberth) [inline lib/trivial.nix:713]: This one's still red, which I think is appropriate because it is fatal.

Comment (Aleksanaa) [reply lib/trivial.nix:713]: Yeah, fair

**Review by infinisil — CHANGES_REQUESTED** [review]
Implementation is looking good to me, but the docs of the function need to be updated a bit.

**Comment by roberth** [comment]
> docs

Done:

   - Bring up to date
   - Give meaning to the metavariables
   - Use italics for metavariables (just like the Nix manual)
   - Don't abbreviate
     - No hard feelings, Val.

**Review by infinisil — APPROVED** [review]
Looking good!

**Comment by roberth** [comment]
Thanks!

**Comment by infinisil** [comment]
Weird to see the "Delete branch" button on somebody else's PR, instinctively clicked it, not sure if you want to keep it around :P

**Pull request merged**